cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java

```diff
@@ -117,53 +120,68 @@
   private final Supplier<WorkflowDataset> workflows;
   private final Supplier<TransactionExecutor> appsTx;
   private final Supplier<TransactionExecutor> workflowsTx;
+  private final MultiThreadDatasetCache dsCache;
 
   @Inject
   public DefaultStore(CConfiguration conf,
                       LocationFactory locationFactory,
                       NamespacedLocationFactory namespacedLocationFactory,
                       final TransactionExecutorFactory txExecutorFactory,
-                      DatasetFramework framework) {
+                      DatasetFramework framework,
+                      TransactionSystemClient txClient) {
     this.configuration = conf;
     this.locationFactory = locationFactory;
     this.namespacedLocationFactory = namespacedLocationFactory;
     this.dsFramework = framework;
-    this.apps = Suppliers.memoize(
+    this.dsCache = new MultiThreadDatasetCache(
+      new SystemDatasetInstantiator(framework, null, null), txClient,
+      Id.Namespace.SYSTEM, ImmutableMap.<String, String>of(), null, null);
+    this.apps =
       new Supplier<AppMetadataStore>() {
         @Override
         public AppMetadataStore get() {
+          Table table;
```

Comment:
Can you turn this get or create table logic into a private method to reduce code duplication?